src/event.rs

```diff
@@ -53,6 +54,30 @@ impl Event {
         unsafe { ffi::gdk_event_put(self.to_glib_none().0) }
     }
 
+    /// Set the event handler.
+    /// 
+    /// The callback `handler` is called for each event. If `None`, event
+    /// handling is disabled.
+    /// 
+    /// This function is marked unsafe since it is a static function with no
+    /// protection against re-entry. It may be called by any thread but must
+    /// not be called concurrently with itself or with either `gtk::init` or
+    /// `gtk::Application::new`.
+    pub unsafe fn set_handler<F: Fn(&mut Event) + Send +'static>(handler: Option<F>) {
+        if let Some(handler) = handler {
```

Comment:
Still need `assert_initialized_main_thread!();` in first line,
this also fix some thread safety problem.

Reply:
Not necessary; the GDK C code is simple, does only one read (from a null-initialised pointer) and is safe to call from any thread (though not reentrant):
```c
void 
gdk_event_handler_set (GdkEventFunc   func,
		       gpointer       data,
		       GDestroyNotify notify)
{
  if (_gdk_event_notify)
    (*_gdk_event_notify) (_gdk_event_data);

  _gdk_event_func = func;
  _gdk_event_data = data;
  _gdk_event_notify = notify;
}
```

Reply:
Then your code don't pass CI 😉

Reply:
Or you can use other macro, `skip_assert_initialized!();` if you sure about no problem in rust,
but IMHO better allow change only from main thread.

Reply:
Yes you can remove the `Send` requirement then.

I meant this one here specifically: https://github.com/gtk-rs/gtk/blob/b17cc04141daeb47971040a58a3d90af42d82c00/src/rt.rs#L59-L62
I don't think the GDK one is actually checking for the main thread